runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -242,15 +252,36 @@ void doSerializeInternal(ByteBuf buf) {
             if (data == null) {
                 int lengthIndex = buf.writerIndex();
                 buf.writeInt(0);
-                Serializers.CORFU.serialize(payload.get(), buf);
+                if (hasPayloadCodec()) {
+                    // if the payload has a codec we need to also compress the payload
+                    ByteBuf serializeBuf =  Unpooled.buffer();
+                    Serializers.CORFU.serialize(payload.get(), serializeBuf);
+                    ByteBuffer wrappedByteBuf = ByteBuffer.wrap(serializeBuf.array(), 0, serializeBuf.readableBytes());
+                    ByteBuffer compressedBuf = getPayloadCodecType().getInstance().compress(wrappedByteBuf);
+                    ICorfuPayload.serialize(buf, Unpooled.wrappedBuffer(compressedBuf));
+                    // Since the server uses the same serialization/deserialization path
+                    // we need a flag to let the server not re-encode an already encoded payload
+                    setCompressedFlag();
+                } else {
+                    Serializers.CORFU.serialize(payload.get(), buf);
+                }
                 int size = buf.writerIndex() - (lengthIndex + 4);
                 buf.writerIndex(lengthIndex);
                 buf.writeInt(size);
                 buf.writerIndex(lengthIndex + size + 4);
             } else {
-                ICorfuPayload.serialize(buf, data);
+                if (hasPayloadCodec() && !isCompressed()) {
+                    // Since the server uses the same serialization/deserialization path
+                    // we need a flag to let the server not re-encode an already encoded payload
```

Comment:
IMO it would be cleaner if you made compression idempotent instead of having these special checks.